doc/developer/design/20220511_replicated_compute_introspection.md

```diff
@@ -0,0 +1,360 @@
+keep them in around for two reasons:
+
+  * Because persist compaction is not implemented yet, introspection storage
+    collections can grow without bounds. We therefore don't want to enable
+    sinking to persist for long-running Materialize deployments (like cloud). To
+    still be able to collect introspection data for such deployments, we need to
+    keep the existing mechanism of peeking introspection arrangements around
+    (even if it only works reliably for single-replica clusters).
+  * Once we have sinks working for replicated clusters, we'll want to think
+    about using them to sink the introspection arrangements instead of the
+    hand-rolled solution proposed in this design.
+
+To sink introspection data, replicas need the following storage connection
+information:
+
+  * Blob URI
+  * Consensus URI
+  * A shard ID for every log variant
+
+This connection information is passed to replicas through the
+`CreateInstance` compute command.
+
+#### Changes to the Compute API
+
+The `LoggingConfig` contained in `ComputeCommand::CreateInstance` is extended to
+also contain the storage connection information mentioned above (details subject
+to bike-shedding):
+
+```rust
+pub struct LoggingConfig {
+    pub granularity_ns: u128,
+    pub active_logs: HashMap<LogVariant, ActiveLog>,
+    pub log_logging: bool,
+    pub blob_uri: String,
+    pub consensus_uri: String,
+}
+
+pub struct ActiveLog {
+    pub index_id: GlobalId,
+    pub shard_id: ShardId,
+}
+```
+
+The compute controller generally assumes that it can send the same commands to
+all replicas in a cluster. This will not be the case anymore with the updated
+`LoggingConfig`, since shard IDs are specific to replicas. Therefore, the
+compute controller must take care of adjusting `CreateInstance` commands to the
+respective target replicas before sending them.
+
+#### Cleanup on Replica Restart
+
+When a replica is restarted for whatever reason, we need to make sure that the
+persist shards they log their introspection data to are cleared. Otherwise,
+we might end up with stale introspection data that was written by a previous
+incarnation of the replica and never retracted because subsequent incarnations
+were not aware of the stale data's existence.
+
+Replicas should clear their introspection shards as soon as they receive a
+`CreateInstance` command, before they initialize the introspection dataflows.
+
+`persist` does not offer a convenient API to clear shards, but it should
+still provide all necessary means. To clear a shard, we can:
+
+  1. Make a snapshot at the current time.
+  1. Issue retraction updates for each record in the snapshot.
+
+Note that this way of clearing shards in-place will not work if an old
+incarnation of the replica is still around logging new data to the same shard,
+as the presence of multiple writers can lead to inconsistent/undefined shard
+contents. This issue will be resolved in the future by the use of storage
```

Comment:
dunno if this is helpful (feel free to ignore if it's not), but the `WriteHandle::compare_and_append` method would let you find out when there's contention on the shard. unexpected contention e.g. could be used as a hint to the old replica that it should go away

Reply:
I assume it would be difficult for a replica to figure out if it is the old or the new one. But this is still valuable, at least we can produce error logs, or panic, to indicate that something is wrong.

Reply:
I don't have a lot of detail on the context, so this could be entirely misguided but perhaps you could use the state (in the state machine sense) of the replica to determine old vs new? e.g. if the protocol is retract the old records followed by indefinitely write new ones...then contention in the first state means you're the new replica and contention in the second state means you're the old one

Reply:
> I don't have a lot of detail on the context, so this could be entirely misguided but perhaps you could use the state (in the state machine sense) of the replica to determine old vs new? e.g. if the protocol is retract the old records followed by indefinitely write new ones...then contention in the first state means you're the new replica and contention in the second state means you're the old one

That would work as a heuristics, but is it guaranteed to always work? I wonder if we could end up with a situation where the old incarnation stops writing just long enough for the new incarnation to finish the "retract state" without seeing any contention. Then when we observe contention, both incarnations are already in the second state, so both believe they are the old one and stop writing.

> You could use an epoch-based approach to this; the shard could track the highest replica id that's written to it and disallow any replica with a lower ID to write, i.e. akin to [distributed locking](https://martin.kleppmann.com/2016/02/08/how-to-do-distributed-locking.html). Seems like a generally desirable feature to let a shard enforce that there is only one writer at a time.

That sounds reasonable to me, but would be something persist needs to support, I think. Any opinions on that @danhhz?